### ISSUE TITLE
Docker deployment

### ISSUE THREAD

**Issue by ChristianGruen** [BaseXdb/basex#2051]
Based on our recent experiences, we’ll simplify and revise our Docker deployment:

Current docker file:
https://github.com/BaseXdb/basex/blob/master/Dockerfile

Current documentation:
https://docs.basex.org/wiki/Docker

Current container:
https://hub.docker.com/r/basex/basexhttp

**Comment by micheee** [comment]
The BaseX Admin Password must be set on first starting the container: 

```
docker run -e BASEX_ADMIN_PASSWORD='123' basex/basexhttp
```

**Comment by hutchkintoot** [comment]
With the new Apple Silicon computers being more and more popular. It would be nice if the image was built with multi-arch support, https://docs.docker.com/desktop/multi-arch/ . 

I have tried building with multi-arch support locally on my intel based Mac, but it fails since the current Dockerfile builds the project with maven inside the image and emulation of arm64 seems to be tricky when it comes to JAVA.

/Johan

**Comment by hutchkintoot** [comment]
Building an image from master changing the base image, I managed to get multi-arch build to work.

### Changed Dockerfile:
``` 
FROM maven:3.5.3-jdk-8
LABEL maintainer="BaseX Team <basex-talk@mailman.uni-konstanz.de>"

# Compile BaseX, install
COPY . /usr/src/basex/

# install git as "buildnumber-maven-plugin" requires git:
RUN apt-get update && apt-get install git && \
    cd /usr/src/basex && \
    mvn clean install -DskipTests && \
    ln -s /usr/src/basex/basex-*/etc/* /usr/local/bin &&\
    adduser --home /srv --disabled-password --uid 1984 basex && \
    mkdir -p /srv/.m2 /srv/basex/data /srv/basex/repo /srv/basex/webapp && \
    cp -r /usr/src/basex/basex-api/src/main/webapp/WEB-INF /srv/basex/webapp && \
    chown -R basex /srv
USER basex
ENV MAVEN_CONFIG=/srv/.m2

# 1984/tcp: API
# 8984/tcp: HTTP
# 8985/tcp: HTTP stop
EXPOSE 1984 8984 8985
VOLUME ["/srv/basex/data", "/srv/basex/repo","/srv/basex/webapp"]
WORKDIR /srv

# Run BaseX HTTP server by default
CMD ["/usr/local/bin/basexhttp"]

```

Build command for multi-arch:
`docker buildx build -f Dockerfile --platform linux/arm64,linux/amd64 --tag basex-multi-arch:latest .`

**Comment by mrdziuban** [comment]
I'd like to help with this if I can! I run basex in docker and would love to be able to upgrade past version 9.5.2 🙂

I'm also developing on an Apple silicon (M1) Mac and @hutchkintoot's Dockerfile above works, but it blows up the image size from 249MB to 749MB since it's no longer using alpine.

As an alternative, I have a working multi-stage Dockerfile that keeps the image size to 264MB, but I'm not sure if there's any special kind of testing I should do to verify everything is working the same as the currently built docker image. Would it make sense for me to open a pull request with my changes and continue the discussion there?

**Comment by apb2006** [comment]
I had a go at this last year, with a focus on small and multi-architechure images see https://github.com/Quodatum/basex-docker
I found, after a bit of playing with docker, I have become quite happy to write my own Dockerfiles to build images as I wanted. 
So I appreciate @micheee short and simple approach.  I think the most useful thing I got out of it was how to use GitHub workflows. This[ build.yml](https://github.com/Quodatum/basex-docker/blob/main/.github/workflows/buildx.yml) publishes to github and docker hub.

**Comment by ChristianGruen** [comment]
I agree with @micheee and @apb2006 it’s usually the best solution to write custom Dockerfiles.

We’ve added a reference to the mailing list on https://hub.docker.com/r/basex/basexhttp and we have (for now) dropped the Docker documentation from our Wiki.

**Comment by ChristianGruen** [comment]
For further discussion, we refer to the mailing list (https://basex.org/about/open-source/).

If anyone is interested in writing an article on BaseX and Docker for our Wiki, don’t hesitate to let us know.

**Comment by timrobertson100** [comment]
Has this been moved please?
https://hub.docker.com/r/basex/basexhttp

**Comment by micheee** [comment]
Hey @timrobertson100  sorry for the inconvenience. 

We decided to rework our outdated Docker releases, we will come back here with an update. 

Until then you have two options, one is using @apb2006 Andy's well maintained images: 

> [3] https://github.com/Quodatum/basex-docker


Apart from that,, if you want to control it fully by yourself, a way to launch a container containing BaseX might be as follows:


## Dockerfile 

```Dockerfile
FROM alpine:3.22 AS fetch

ARG BASEX_VERSION=12.4

RUN apk add --no-cache unzip \
    && wget -q "https://files.basex.org/releases/${BASEX_VERSION}/BaseX$(echo "${BASEX_VERSION}" | tr -d '.').zip" -O /tmp/basex.zip \
    && unzip -q /tmp/basex.zip -d /opt

FROM eclipse-temurin:21-jre-alpine

LABEL org.opencontainers.image.source="https://basex.org"

RUN apk add --no-cache bash \
    && addgroup -S basex && adduser -S basex -G basex

COPY --from=fetch --chown=basex:basex /opt/basex /opt/basex
COPY --chmod=755 docker-entrypoint.sh /usr/local/bin/

ENV PATH=/opt/basex/bin:$PATH

WORKDIR /opt/basex

VOLUME ["/opt/basex/data"]

USER basex

# 1984: database server
# 8080: HTTP
EXPOSE 8080

# / is served by webapp/restxq.xqm
HEALTHCHECK --interval=30s --timeout=5s --start-period=15s --retries=3 \
    CMD wget -q -O /dev/null http://localhost:8080/

# Sets the admin password on first run (see docker-entrypoint.sh).
ENTRYPOINT ["docker-entrypoint.sh"]
CMD ["basexhttp"]

```


## docker-entrypoint.sh to set a password on first run

```bash
#!/bin/bash
# file: docker-entrypoint.sh
set -e

# On first run (no users.xml yet), set the admin password: either the one given
# via BASEX_ADMIN_PASSWORD, or a generated one that is printed to the log.
if [ ! -f /opt/basex/data/users.xml ]; then
  if [ -n "$BASEX_ADMIN_PASSWORD" ]; then
    basex -c "PASSWORD $BASEX_ADMIN_PASSWORD"
  else
    PW="$(tr -dc 'A-Za-z0-9' < /dev/urandom | head -c 16)"
    basex -c "PASSWORD $PW"
    echo "Generated admin password: $PW"
  fi
fi

exec "$@"
```

Best
Michael

**Comment by timrobertson100** [comment]
Thanks for replying [micheee](https://github.com/micheee) - good to know.
My colleague (@MattBlissett) is working up a replacement dockerfile for our needs.

**Comment by MattBlissett** [comment]
Thanks all — I'm now back up and running with a custom Dockerfile.

**Comment by ChristianGruen** [comment]
Merged: https://github.com/BaseXdb/basex/pull/2713

**Comment by ChristianGruen** [comment]
The documentation has been updated: https://docs.basex.org/13/Docker
